lsp/nls/src/requests/completion.rs

```diff
@@ -167,33 +168,72 @@ fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
 /// partiular ID, and in the scope of a given linearization item.
 fn collect_record_info(
     linearization: &Completed,
-    item: &LinearizationItem<Types>,
     id: usize,
     path: &mut Vec<Ident>,
 ) -> Vec<(Vec<Ident>, Types)> {
-    linearization
-        .get_in_scope(item)
-        .iter()
-        .filter_map(|item| {
-            let (ty, _) = linearization.resolve_item_type_meta(item);
-            match (&item.kind, ty) {
-                // Get record fields from static type info
-                (TermKind::Declaration(..), Types(TypeF::Record(rrows))) if id == item.id => {
-                    Some((find_fields_from_type(&rrows, path), item.ty.clone()))
-                }
-                (TermKind::Declaration(_, _, ValueState::Known(body_id)), _) if id == item.id => {
-                    match find_fields_from_contract(&linearization, *body_id, path) {
-                        // Get record fields from contract metadata
-                        Some(fields) => Some((fields, item.ty.clone())),
-                        // Get record fields from lexical scoping
-                        None => find_fields_from_term_kind(&linearization, id, path)
-                            .map(|idents| (idents, item.ty.clone())),
+    if let Some(item) = linearization.get_item(id) {
+        let (ty, _) = linearization.resolve_item_type_meta(item);
+        match (&item.kind, ty) {
+            // Get record fields from static type info
+            (TermKind::Declaration(..), Types(TypeF::Record(row))) => {
+                vec![(find_fields_from_type(&row, path), item.ty.clone())]
+            }
+            (TermKind::Declaration(_, _, ValueState::Known(body_id)), _) => {
+                match find_fields_from_contract(&linearization, *body_id, path) {
+                    // Get record fields from contract metadata
+                    Some(fields) => vec![(fields, item.ty.clone())],
+                    // Get record fields from lexical scoping
+                    None => {
+                        if let Some(idents) = find_fields_from_term_kind(&linearization, id, path) {
+                            vec![(idents, item.ty.clone())]
+                        } else {
+                            return Vec::new();
+                        }
                     }
```

Comment:
I think you could probably avoid pattern-matching on the optional here, with something like:

```suggestion
                find_fields_from_contract(&linearization, *body_id, path)
                    .or_else(|| find_fields_from_term_kind(&linearization, id, path))
                    .map(|idents| vec![(idents, item.ty.clone())])
                    .unwrap_or_default()
```